### DISCUSSION TITLE
Using ref and Controller

### DISCUSSION THREAD

**Discussion by cwiese** [react-hook-form/react-hook-form#3565] [Q&A] [answered]
I am going in circles.  Trying to get Material-UI, react-number-format working.  Updated to latest version of react-hook-form. 

I added `ref={register({ valueAsNumber: true, required: true })}` to NumberFormat in an attempt to get the value to update onSubmit as a number not a string ... ref={register} does not complain.  However, when I add validation or "valueAsNumber" i get "No overload matches this call. Overload 1 of 2, '(props: NumberFormatProps"

Any help appreciated. 

`import { Controller, useForm } from 'react-hook-form';
import Input from '@material-ui/core/Input';
import NumberFormat from "react-number-format";

      <Controller
        as={
          <NumberFormat
            label="Averaging"
            thousandSeparator={true}
            customInput={TextField}
            inputRef={register({ valueAsNumber: true, required: true })}
          />
        }
        defaultValue={params.avg.toString()}
        name="avg"
        variant="outlined"
        control={control}
        mode="onBlur"
        error={!!errors.avg}
        helperText={errors.avg ? errors.avg.message : ''}
      />`

**Comment by bluebill1049** [comment, marked as answer]
Controller will handle the register, if you need to convert type with Controller have a look at here: https://react-hook-form.com/advanced-usage#TransformandParse

**Reply by husk-9** [reply]
Is there not simpler way of doing this? 

In my project, I have various components including `Input` and `Select`, and I am trying to make them `react-hook-form` components. `valueAsNumber` and `valueAsDate` are not compatible with Controllers. I thought perhaps if I make a form using a yup schema resolver, it will coerce the values to date/number where appropriate, but it seems that this is not the case.

The transform and parse way certainly works, but it seems like a lot of extra work to make sure that my custom components are returning numbers as `4` and not `"4"`. Am I missing some documentation on controllers?

**Reply by bluebill1049** [reply]
```tsx
<Controller render={({ field }) => {
  return <input value={+field} onChange={(e) => field.onChange(+e.target.value)} />
}} />
``` 
what stop you from doing this above? controlled component can change value to any formate that you want, controller's job is only to collect data and validate.

**Comment by cwiese** [comment]
okay thanks (thanks for the quick reply) ..... how about the "required: true"?  do I need to switch to a validation resolver ?

**Reply by bluebill1049** [reply]
you can use the `rules` prop

**Comment by cwiese** [comment]
I used that example (which cleans up boilerplate code also) ... even though I can see the console.log of the number value entered, oddly, now the value saved to state is the default value.    Even if I remove onChange the changed value is not getting used.

(having issues formatting code)

const ControllerNum = ({
    control, name, defaultValue, errors, label, register
}: {control: any, name: string, defaultValue: number, errors: any, label: string, register: any}) => (
  <Controller
    defaultValue={defaultValue}
    control={control}
    name={name}
    variant="outlined"
    render={(props) => (
      <NumberFormat
        label={label}
        thousandSeparator={true}
        customInput={TextField}
        onChange={(e) => {
          const output = parseFloat(e.target.value!);
          console.log("rtn2",output)
          return output;
        }}
        value={isNaN(props.value) ? "" : props.value.toString()}
    />)}
    error={!!errors[name]}
    helperText={errors[name] ? errors[name].message : ''}
  />
);


used in a form :

`    <ControllerNum
        name="avg"
        label="Averaging"
        defaultValue={params.avg}
        control={control}
        errors={errors}
        register={register}
      />`

**Reply by bluebill1049** [reply]
you didn't send any value into `Contorller`

`props.onChange(//value here)`

**Reply by cwiese** [reply]
missed that -- now it works!  Thank you

**Reply by bluebill1049** [reply]
np

**Comment by cwiese** [comment]
one last thing ... even though the wrapped Controller (ControllerNum above) has    'variant="outlined"' it is showing up as underlined ....only happens with the wrapped version.

**Comment by cwiese** [comment]
![image](https://user-images.githubusercontent.com/178527/100559249-d45c8280-327f-11eb-82a6-65c045d5c970.png)

**Reply by cwiese** [reply]
they all have variant="outlined"

**Comment by S0PEX** [comment]
Is there any rule of thumb when to use Controllers  or register?
It looks like `<Controller>` are handy when wrapping controlled Input. Otherwise it looks like `ref={register()} `.

Is this sort of right or what's the best way?
